Source/DafnyStandardLibraries/src/Std/Parsers/Core/ParsersBuilders.dfy

```diff
@@ -0,0 +1,501 @@
+// B(p)       returns a parser builder from a normal parser.
+// B1.e_I(B2) will parse both but return the result of B2
+// B1.I_e(B2) will parse both but return the result of B1
+// B.M(f)     will map the result of the parser builder by f if succeeded
+// B1.O(B2)   will either parse B1, or B2 if B1 fails with Recoverable
+// O([B1, B2, B3])
+//            will parse with B1, but if B1 fails with Recoverable,
+//            it will parse with B2, but if B2 fails with Recoverable,
+//            it will parse with B3
+// R(v)       returns a parser builder that returns immediately v
+
+abstract module Std.Parsers.Builders {
+  import P: Core
+  export
+    provides P
+    provides O
+    provides SucceedWith
+    provides FailWith
+    provides ResultWith
+    provides Rec
+    provides MId
+    provides B.Apply
+    provides B.e_I
+    provides B.I_e
+    provides B.I_I
+    provides B.M, B.M2, B.M3
+    provides B.If
+    provides B.?
+    provides B.??
+    provides B.Then
+    provides B.ThenWithRemaining
+    provides B.Bind
+    provides B.Rep
+    provides B.RepFold
+    provides B.RepSep
+    provides B.RepMerge
+    provides B.RepSepMerge
+    provides B.Rep1
+    provides B.Debug
+    provides B.End
+    provides EOS
+    provides Nothing
+    provides CharTest
+    provides ToInput
+    provides RecNoStack
+    provides RecursiveProgressError
+    reveals NonProgressing
+    reveals Input
+    reveals InputLength
+    reveals C
+    reveals ParseResult
+    reveals B
+    reveals Rec, RecMap
+    reveals RecMapDef, FailureLevel, RecMapSel
+    reveals RecNoStackResult
+
+  type Input = P.Input
+  type C = P.C
+
+  /** `FailureLevel` is an alias for `P.FailureLevel`. Recoverable or Fatal */
+  type FailureLevel = P.FailureLevel
+  /** A mapping from string to a parser returning `A` */
+  type RecMapSel<A> = string -> B<A>
+  /** `ParseResult` is an alias for `P.ParseResult`. */
+  type ParseResult<+T> = P.ParseResult<T>
+
+  function ToInput(other: seq<C>): (i: Input)
+
+  /** `SucceedWith(a)` returns success on any input,
+      `.value` contains the provided value.*/
+  function SucceedWith<T>(t: T): B<T> {
+    B(P.SucceedWith(t))
+  }
+
+  /** `FailWith(message)` returns a failure on any input, and does not consume input. */
+  function FailWith<T>(message: string, level: FailureLevel := P.Recoverable): B<T> {
+    B(P.FailWith(message,  level))
+  }
+
+  /** A parser that always returns the given result  */
+  function ResultWith<R>(result: ParseResult<R>): B<R>
+  {
+    B(P.ResultWith(result))
+  }
+
+  /** Parses nothing, always returns a success with nothing consumed. Can be useful in alternatives. */
+  const Nothing: B<()> := B(P.Epsilon())
+
+  /** An identity function useful for B.M2 and B.M3 */
+  function MId<R>(r: R): R { r }
+
+  /** Adapter to wrap any underlying parser so that they can be chained together. */
+  datatype B<R> = B(apply: P.Parser<R>)
```

Comment:
We can call this `Builder` right? I'm assuming it usually only shows up at the beginning of the builder chain?

Reply:
See [this comment](https://github.com/dafny-lang/dafny/pull/6143/#discussion_r2083115602) as I'm adding a description about why short names in parser combinators matter.
This includes the name for builders itself, which repetitively appears in lambda expressions when chaining parsers.